src/pam.rs

```diff
@@ -10,30 +10,64 @@ use std::str::Utf8Error;
 pub struct Pam(PamHandle);
 
 impl Pam {
+
+    /// Get the username. If the PAM_USER item is not set, this function
+    /// prompts for a username (like get_authtok).
+    pub fn get_user(&self, prompt: Option<&str>) -> PamResult<Option<&str>> {
+        let cprompt = prompt.map(|p| CString::new(p).expect("Error, the prompt cannot contain any null bytes"));
+        let pointer = get_user(self.0, cprompt.as_ref().map(|p| p.as_ptr()))?;
+        ptr_to_str(pointer)
+    }
+
+    /// Get the username, i.e. the PAM_USER item. If it's not set return None.
+    pub fn get_cached_user(&self) -> PamResult<Option<&str>> {
+        let pointer = get_item(self.0, PamItemType::AUTHTOK)?;
+        ptr_to_str(pointer)
+    }
+
     /// Get the cached authentication token.
-    pub fn get_cached_authtok(&self) -> PamResult<Option<&CStr>> {
-        // pam should keep the underlying token allocated for as long as the module is loaded
-        // which make this safe
+    pub fn get_cached_authtok(&self) -> PamResult<Option<&[u8]>> {
         unsafe {
             let pointer = try!(get_item(self.0, PamItemType::AUTHTOK));
-            Ok(pointer.map(|p| CStr::from_ptr(p as *const c_char)))
+            Ok(pointer.map(|p| CStr::from_ptr(p as *const c_char).to_bytes()))
         }
     }
 
     /// Get the cached authentication token or prompt the user for one if there isn't any
-    pub fn get_authtok(&self, prompt: Option<&str>) -> PamResult<Option<&CStr>> {
+    pub fn get_authtok(&self, prompt: Option<&str>) -> PamResult<Option<&[u8]>> {
         let cprompt = prompt.map(|p| CString::new(p).expect("Error, the prompt cannot contain any null bytes"));
         let result = try!(get_authtok(self.0, PamItemType::AUTHTOK, cprompt.as_ref().map(|p| p.as_ptr())));
         // If result is Ok we're guaranteed that p is a valid pointer
         unsafe {
-            Ok(result.map(|p| CStr::from_ptr(p)))
+            Ok(result.map(|p| CStr::from_ptr(p).to_bytes()))
         }
     }
 
-    pub fn set_authtok(&self, authtok: &CString) -> PamResult<()> {
+    pub fn set_authtok(&self, authtok: &[u8]) -> PamResult<()> {
         set_item(self.0, PamItemType::AUTHTOK, authtok.as_ptr() as *const c_void)
     }
 
+    /// Get the remote hostname.
+    pub fn get_rhost<'a>(&'a self) -> PamResult<Option<&'a str>> {
+        let pointer = get_item(self.0, PamItemType::RHOST)?;
+        ptr_to_str(pointer)
+    }
+
+    /// Get the remote username.
+    pub fn get_ruser<'a>(&'a self) -> PamResult<Option<&'a str>> {
+        let pointer = get_item(self.0, PamItemType::RUSER)?;
+        ptr_to_str(pointer)
+    }
+}
+
+fn ptr_to_str(pointer: Option<*const c_void>) -> PamResult<Option<&'static str>> {
```

Comment:
I did that so that a module writer would not have to write any unsafe code. Note that the username is &str, but the password is &[u8] for that exact reason. We could change username to &[u8] as well.

A more interesting issue is, should we return an unsafe &Cstr and force the module author to use an unsafe{} block? If not, and we return a (safe) &[u8] slice instead, how safe _is_ that - I'm not quite sure about the lifetime of that data. There's always the option of simply returning owned data, a Vec<u8>, but that means an extra allocation.

I don't know. Please make a decision and I'll write the code :)

Reply:
Okay, I took the time to read the CString/CStr source code from the library, because it's clear I did not understand how CStr worked. CStr is just a [c_char], and &CStr is not a reference to some struct, but a slice (like &str). I just reasoned about it in the wrong way, TIL :). I do agree that returning a &CStr is the way to go.